src/Sulu/Bundle/AdminBundle/Resources/js/containers/FieldBlocks/FieldBlocks.js

```diff
@@ -13,6 +13,46 @@ const MISSING_BLOCK_ERROR_MESSAGE = 'The "block" field type needs at least one t
 const BLOCK_PREVIEW_TAG = 'sulu.block_preview';
 
 export default class FieldBlocks extends React.Component<FieldTypeProps<Array<BlockEntry>>> {
+    componentDidUpdate(prevProps: FieldTypeProps<Array<BlockEntry>>) {
+        const {defaultType, types, value: oldValues, onChange} = this.props;
+        const {types: oldTypes} = prevProps;
+
+        if (!oldValues) {
+            return;
+        }
+
+        if (!types || !oldTypes) {
+            throw new Error(MISSING_BLOCK_ERROR_MESSAGE);
+        }
+
+        // do nothing when types is same reference as oldTypes
+        if (types === oldTypes) {
+            return;
+        }
+
+        const newValues = toJS(oldValues);
+        const usedTypes = newValues.map((block) => block.type);
+
+        // do nothing when used types where included in oldTypes
+        if (usedTypes.every((type) => Object.keys(oldTypes).includes(type))) {
+            return;
+        }
```

Comment:
@danrot this could be omited by doing something like beneath:

```
        let hasChanged = false;

        newValues.forEach((blockValue, i) => {
            if (!types[blockValue.type]) {
                hasChanged = true;
                newValues[i].type = defaultType;
            }
        });

        if (hasChanged) {
            onChange(newValues)
        }
```

what do you like better?

Reply:
Personally I prefer solutions without `let` :see_no_evil: And as written above I would turn that statement around, to avoid a early return in `componentDidUpdate`:

```javascript
if (!usedTypes.every((type) => Object.keys(oldTypes).includes(type))) {
    // Do all the required stuff when types are not available
}
```